Update models.md (#6331)
pydantic/pydantic: docs/usage/models.md

The primary means of defining objects in Pydantic is via models. Models are simply classes which inherit from

> `pydantic.BaseModel`.

Reviewer: ```suggestion
[`pydantic.BaseModel`][pydantic.BaseModel].
```

You can think of models as similar to structs in languages like C, or as the requirements of a single endpoint
in an API.

> Models share many similarities with Python's dataclasses, but have been designed with some subtle but important

Reviewer: ```suggestion
Models share many similarities with Python's dataclasses, but have been designed with some subtle-yet-important
```

    The example above works because aliases have priority over field names for
    field population. Accessing `SQLModel`'s `metadata` attribute would lead to a `ValidationError`.

### Nested attributes

> When using attributes to parse models, Model instances will be created from both top level attributes and

Reviewer: ```suggestion
When using attributes to parse models, model instances will be created from both top-level attributes and
```

Pydantic will raise `ValidationError` whenever it finds an error in the data it's validating.

A single exception of type `ValidationError` will be raised regardless of the number of errors found,

> and that `ValidationError` will contain information about all the errors and how they happened.

Reviewer: ```suggestion
and that `ValidationError` will contain information about all of the errors and how they happened.
```

    """
```

> If you want to validate serialized data in a format besides JSON, you should load the data into a dict yourself and

Reviewer: ```suggestion
If you want to validate serialized data in a format other than JSON, you should load the data into a dict yourself and
```

    """
```

> If you set the `model_config` or make use of `@field_validator` or other pydantic decorators in your generic model

Reviewer: ```suggestion
If you set the `model_config` or make use of `@field_validator` or other Pydantic decorators in your generic model
```

definition, they will be applied to parametrized subclasses in the same way as when inheriting from a `BaseModel`
subclass. Any methods defined on your generic class will also be inherited.

Pydantic's generics also integrate properly with type checkers, so you get all the type checking
you would expect if you were to declare a distinct type for each parametrization.

!!! note

>     Internally, pydantic creates subclasses of `BaseModel` at runtime when generic models are parametrized.

Reviewer: ```suggestion
    Internally, Pydantic creates subclasses of `BaseModel` at runtime when generic models are parametrized.
```

    """
```

When using bound type parameters, and when leaving type parameters unspecified, Pydantic treats generic models
similarly to how it treats built-in generic types like `List` and `Dict`:

> * If you don't specify parameters before instantiating the generic model, they are treated as the bound of the `TypeVar`

Reviewer: ```suggestion
* If you don't specify parameters before instantiating the generic model, they are treated as the bound of the `TypeVar`.
```

> * If the `TypeVar`s involved have no bounds, they are treated as `Any`

Reviewer: ```suggestion
* If the `TypeVar`s involved have no bounds, they are treated as `Any`.
```

## `RootModel` and custom root types

> Pydantic models can be defined with a "custom root type" by subclassing `pydantic.RootModel`.

Reviewer: ```suggestion
Pydantic models can be defined with a "custom root type" by subclassing [`pydantic.RootModel`][pydantic.RootModel].
```

> The root type can be any type supported by pydantic, and is specified by the generic parameter to `RootModel`.

Reviewer: ```suggestion
The root type can be any type supported by Pydantic, and is specified by the generic parameter to `RootModel`.
```

#> ['dog', 'cat']
```

You can also create subclasses of the parametrized root model directly:

```py
from typing import List

from pydantic import RootModel

class Pets(RootModel[List[str]]):
    root: List[str]

    def describe(self) -> str:
        return f'Pets: {", ".join(self.root)}'

my_pets = Pets.model_validate(['dog', 'cat'])

print(my_pets.describe())
#> Pets: dog, cat
```

## Faux immutability

Models can be configured to be immutable via `model_config['frozen'] = True`. When this is set, attempting to change the

> values of instance attributes will raise errors. See [model config](model_config.md) for more details on `Config`.

Reviewer: ```suggestion
values of instance attributes will raise errors. See [Model Config](model_config.md) for more details.
```

assert m1.updated != m2.updated
```

You can find more information in the documentation of the [`Field` function](fields.md).

## Automatically excluded attributes

### Class vars

> Attributes annotated with `typing.ClassVar` will are properly treated by Pydantic as class variables, and will not

Reviewer: ```suggestion
Attributes annotated with `typing.ClassVar` are properly treated by Pydantic as class variables, and will not
```

```

To be included in the signature, a field's alias or name must be a valid Python identifier.
Pydantic will prioritize a field's alias over its name when generating the signature, but may use the field name if the
alias is not a valid Python identifier.

> If a field's alias and name are _both_ invalid identifiers (which may be possible through exotic use of `create_model`),

Reviewer: ```suggestion
If a field's alias and name are _both_ not valid identifiers (which may be possible through exotic use of `create_model`),
```
